src/BTrees/_datatypes.py

```diff
@@ -161,58 +242,9 @@ def as_value_type(self):
     def supports_value_union(self):
         return False
 
-    @staticmethod
-    def _check_default_comparison(
-            item,
-            # PyPy2 doesn't define __lt__ on object; PyPy3 and
-            # CPython 2 and 3 do.
-            _object_lt=getattr(object, '__lt__', object())
-    ):
-        # Enforce test that key has non-default comparison.
-        # (With the exception of None, because we define a sort order
-        # for it.)
-
-        if item is None:
-            return
-
-        if type(item) is object: # pylint:disable=unidiomatic-typecheck
-            raise TypeError("Can't use object() as keys")
-
-        # Now more complicated checks to be sure we didn't
-        # inherit default comparison on any version of Python.
-
-
-        # TODO: Using a custom ABC and doing ``isinstance(item, NoDefaultComparison)``
-        # would automatically gain us caching.
-
-        # XXX: Comparisons only use special methods defined on the
-        # type, not instance variables. So shouldn't this be getattr(type(key)...)?
-        # Note that changes some things below; for example, on CPython 2,
-        # every subclass of object inherits __lt__ (unless it overrides), and it
-        # has __objclass__ of ``type`` (of course it is also the same object
-        # as ``_object_lt`` at that point). Also, weirdly, CPython 2 classes inherit
-        # ``__lt__``, but *instances* do not.
-
-        lt = getattr(item, '__lt__', None)
-        if lt is not None:
-            # CPython 2 and 3 follow PEP 252, defining '__objclass__'
-            # for methods of builtin types like str; methods of
-            # classes defined in Python don't get it. ``__objclass__``
-            if getattr(lt, '__objclass__', None) is object:
-                lt = None  # pragma: no cover Py3k
-            # PyPy3 doesn't follow PEP 252, but defines '__func__'
-            elif getattr(lt, '__func__', None) is _object_lt:
-                lt = None  # pragma: no cover PyPy3
-
-        if (lt is None
-                # TODO: Shouldn't we only check __cmp__ on Python 2?
-                # Python 3 won't use it.
-                and getattr(item, '__cmp__', None) is None):
-            raise TypeError("Object has default comparison")
-
-
     def __call__(self, item):
-        self._check_default_comparison(item)
+        if isinstance(item, _HasDefaultComparison):
```

Comment:
`_HasDefaultComparison` is defined as a new class above but where is `item` registered for it, so this check can become true?

Reply:
> _HasDefaultComparison is defined as a new class above

Right. Specifically, it's defined as a new abstract base class, i.e., something that extends `abc.ABCMeta`. (We could do this with a regular `type` subclass, but extending `ABCMeta` gets us caching of the results of `isinstance` automatically.)

> where is `item` registered for it, so this check can become true?

That's the beauty part (as the saying goes). By making `_HasDefaultComparison` implement `__subclasshook__`, we [algorithmically define](https://docs.python.org/3/library/abc.html#abc.ABCMeta.__subclasshook__) what `isinstance(item, _HasDefaultComparison)` means. No need to register anything! The algorithm is the same as what was here before, just moved into `__subclasshook__`. (The expression `isinstance(item, Cls)` basically boils down to `Cls.__subclasscheck__(type(item))`; `ABCMeta` defines `__subclasscheck__` to do some caching, and if the answer is not in the cache, to call `Cls.__subclasshook__()`.)

Reply:
Did that help clear things up sufficiently?

Reply:
TIL, I think I now understood how it works. Maybe adding a link to the Python documentation would help future readers of the code who rarely use abstract base classes.

Reply:
Done!